From ffb25225d1b15e2a102f9a1503d9f17f456e6e77 Mon Sep 17 00:00:00 2001
From: Chad Dupuis <cdupuis@redhat.com>
Date: Thu, 21 Aug 2014 13:05:28 -0400
Subject: [ethernet] qlcnic: Fix TSS/RSS ring validation logic

Message-id: <1408626422-3941-45-git-send-email-cdupuis@redhat.com>
Patchwork-id: 88568
O-Subject: [PATCH RHEL 7.1 044/138] qlcnic: Fix TSS/RSS ring validation logic.
Bugzilla: 1089352
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>

From: Himanshu Madhani <himanshu.madhani@qlogic.com>

Bug 1089352 (https://bugzilla.redhat.com/show_bug.cgi?id=1089352)

net commit id f9566265d7b44fea789072dcfa9a454e7e433af6

o TSS/RSS ring validation does not take into account that either
  of these ring values can be 0. This patch fixes this validation
  and would fail set_channel operation if any of these ring value
  is 0. This regression was added as part of commit id
  34e8c406fda5b5a9d2e126a92bab84cd28e3b5fa ("qlcnic: refactor Tx/SDS
  ring calculation and validation in driver.")

Signed-off-by: Himanshu Madhani <himanshu.madhani@qlogic.com>
Signed-off-by: David S. Miller <davem@davemloft.net>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/net/ethernet/qlogic/qlcnic/qlcnic_83xx_init.c b/drivers/net/ethernet/qlogic/qlcnic/qlcnic_83xx_init.c
index 89208e5..fae1b71 100644
--- a/drivers/net/ethernet/qlogic/qlcnic/qlcnic_83xx_init.c
+++ b/drivers/net/ethernet/qlogic/qlcnic/qlcnic_83xx_init.c
@@ -2073,8 +2073,8 @@ int qlcnic_83xx_configure_opmode(struct qlcnic_adapter *adapter)
   ahw->nic_mode = QLCNIC_DEFAULT_MODE;
   adapter->nic_ops->init_driver = qlcnic_83xx_init_default_driver;
   ahw->idc.state_entry = qlcnic_83xx_idc_ready_state_entry;
-  adapter->max_sds_rings = ahw->max_rx_ques;
-  adapter->max_tx_rings = ahw->max_tx_ques;
+  adapter->max_sds_rings = QLCNIC_MAX_SDS_RINGS;
+  adapter->max_tx_rings = QLCNIC_MAX_TX_RINGS;
  } else {
   return -EIO;
  }
diff --git a/drivers/net/ethernet/qlogic/qlcnic/qlcnic_ethtool.c b/drivers/net/ethernet/qlogic/qlcnic/qlcnic_ethtool.c
index 6d3edf6..78f5e81 100644
--- a/drivers/net/ethernet/qlogic/qlcnic/qlcnic_ethtool.c
+++ b/drivers/net/ethernet/qlogic/qlcnic/qlcnic_ethtool.c
@@ -667,9 +667,13 @@ qlcnic_set_ringparam(struct net_device *dev,
 static int qlcnic_validate_ring_count(struct qlcnic_adapter *adapter,
           u8 rx_ring, u8 tx_ring)
 {
+ if (rx_ring == 0 || tx_ring == 0)
+  return -EINVAL;
+
  if (rx_ring != 0) {
   if (rx_ring > adapter->max_sds_rings) {
-   netdev_err(adapter->netdev, "Invalid ring count, SDS ring count %d should not be greater than max %d driver sds rings.\n",
+   netdev_err(adapter->netdev,
+       "Invalid ring count, SDS ring count %d should not be greater than max %d driver sds rings.\n",
        rx_ring, adapter->max_sds_rings);
    return -EINVAL;
   }
diff --git a/drivers/net/ethernet/qlogic/qlcnic/qlcnic_main.c b/drivers/net/ethernet/qlogic/qlcnic/qlcnic_main.c
index d08f0c0..655d1bb 100644
--- a/drivers/net/ethernet/qlogic/qlcnic/qlcnic_main.c
+++ b/drivers/net/ethernet/qlogic/qlcnic/qlcnic_main.c
@@ -1178,6 +1178,7 @@ qlcnic_initialize_nic(struct qlcnic_adapter *adapter)
  } else {
   adapter->ahw->nic_mode = QLCNIC_DEFAULT_MODE;
   adapter->max_tx_rings = QLCNIC_MAX_HW_TX_RINGS;
+  adapter->max_sds_rings = QLCNIC_MAX_SDS_RINGS;
   adapter->flags &= ~QLCNIC_ESWITCH_ENABLED;
  }
 
-- 
1.7.1